File: Scala/src/main/scala/AlgorithmicHeights/QS.scala
package AlgorithmicHeights

object QS {
  import scala.annotation.tailrec
  import scala.util.Random

  private val rng = new Random(2112)

  private def convertToIntArray(line: String): Array[Int] = line.split(" ").map(_.toInt)

  private def swapElems(array: Array[Int], ix: Int, jy: Int): Unit = {
    if (ix != jy) {
      val elem: Int = array(ix)
      array(ix) = array(jy)
      array(jy) = elem
    }
  }

  private def threeWayPartitioning(array: Array[Int], pivot: Int, startIx: Int, endIx: Int): (Int, Int) = {
    @tailrec
    def loop(currentIx: Int, startIx: Int, endIx: Int): (Int, Int) = {
      if (currentIx > endIx) (startIx, endIx + 1)
      else if (array(currentIx) < pivot) {
        swapElems(array, currentIx, startIx)
        loop(currentIx + 1, startIx + 1, endIx)
      }
      else if (array(currentIx) > pivot) {
        swapElems(array, currentIx, endIx)
        loop(currentIx, startIx, endIx - 1)
      }
      else loop(currentIx + 1, startIx, endIx)
    }

    loop(currentIx = startIx, startIx = startIx, endIx = endIx)
  }

  def quickSort(array: Array[Int], n: Int): Unit = {
    @tailrec
    def loop(stack: List[(Int, Int)]): Unit = stack match {
      case Nil => ()
      case (startIx, endIx) :: rest =>
        if (startIx >= endIx) loop(rest)
        else {
          val randomIx: Int = rng.between(startIx, endIx + 1)
          val pivot: Int = array(randomIx)
          val (middleStart, middleEnd): (Int, Int) = threeWayPartitioning(array, pivot, startIx, endIx)
          loop((startIx, middleStart - 1) :: (middleEnd, endIx) :: rest)
        }
    }
    loop(List((0, n - 1)))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val array: Array[Int] = convertToIntArray(reader.next())
    quickSort(array, n)
    println(array.mkString(" "))
  }
}

File: OCaml/src/algorithmic_heights/qs.ml
let convert_to_ints (line : string) : int array =
    Array.of_list @@ List.map int_of_string Str.(line |> split (regexp " "))


let three_way_partitioning (arr : int array) (pivot : int) (start_ix : int) (end_ix : int) : int * int =
    let swap i j =
        let elem = arr.(i) in
        arr.(i) <- arr.(j);
        arr.(j) <- elem in
    let rec loop current_ix start_ix end_ix =
        if current_ix > end_ix then (start_ix, end_ix + 1)
        else if arr.(current_ix) < pivot then
            (swap current_ix start_ix;
            loop (current_ix + 1) (start_ix + 1) end_ix)
        else if arr.(current_ix) > pivot then
            (swap current_ix end_ix;
            loop current_ix start_ix (end_ix - 1))
        else
            loop (current_ix + 1) start_ix end_ix in
    loop start_ix start_ix end_ix


let quicksort (arr : int array) (n : int) : unit =
    Random.init 2112;
    let rec loop = function
        | [] -> ()
        | (start_ix, end_ix) :: rest ->
            if start_ix >= end_ix then loop rest
            else
                let random_ix = start_ix + Random.int (end_ix - start_ix + 1) in
                let pivot = arr.(random_ix) in
                let middle_start, middle_end = three_way_partitioning arr pivot start_ix end_ix in
                loop ((start_ix, middle_start - 1) :: (middle_end, end_ix) :: rest) in
    loop [(0, n - 1)]


let () =
    let n = read_int () in
    let arr = convert_to_ints (read_line ()) in
    quicksort arr n;
    Array.(arr |> map string_of_int |> to_list) |> String.concat " " |> print_endline
